[Site Isolation] Make unresponsive process handling work

rniwa opened the pull request:
#### 21ec19ce100f4f11702815b94c1f9f1faa59b791
<pre>
[Site Isolation] Make unresponsive process handling work
<a href="https://bugs.webkit.org/show_bug.cgi?id=302842">https://bugs.webkit.org/show_bug.cgi?id=302842</a>

Reviewed by Sihui Liu and Ben Nham.

Make the handling of foreground unresponsiveness work with site isolation.

Test: TestWebKitAPI.SiteIsolation.UnresponsiveProcessKeydown
      TestWebKitAPI.SiteIsolation.ResponsiveProcessAfterMousedown
      TestWebKitAPI.SiteIsolation.UnresponsiveProcessMousedown
      TestWebKitAPI.SiteIsolation.UnresponsiveProcessDies

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::startResponsivenessTimerForMouseEvent):
(WebKit::WebPageProxy::processNextQueuedMouseEvent):
(WebKit::WebPageProxy::processNextQueuedGestureEvent):
(WebKit::WebPageProxy::sendKeyEvent):
(WebKit::WebPageProxy::handleKeyboardEvent):
(WebKit::WebPageProxy::mouseEventHandlingCompleted):
(WebKit::WebPageProxy::didReceiveEvent):
(WebKit::WebPageProxy::processDidBecomeUnresponsive):
(WebKit::WebPageProxy::processDidBecomeResponsive):
(WebKit::WebPageProxy::dispatchProcessDidTerminate):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::removeRemotePageProxy):
(WebKit::WebProcessProxy::removeWebPage):
(WebKit::WebProcessProxy::didBecomeUnresponsive):
(WebKit::WebProcessProxy::didBecomeResponsive):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(-[NavigationDelegateWithUnresponsiveCallback webView:didReceiveAuthenticationChallenge:completionHandler:]):
(-[NavigationDelegateWithUnresponsiveCallback waitForDidFinishNavigation]):
(-[NavigationDelegateWithUnresponsiveCallback didBecomeUnresponsive]):
(-[NavigationDelegateWithUnresponsiveCallback didBecomeResponsive]):
(-[NavigationDelegateWithUnresponsiveCallback _webViewWebProcessDidBecomeUnresponsive:]):
(-[NavigationDelegateWithUnresponsiveCallback _webViewWebProcessDidBecomeResponsive:]):
(-[NavigationDelegateWithUnresponsiveCallback webView:didFinishNavigation:]):
(TestWebKitAPI::TEST(SiteIsolation, UnresponsiveProcessKeydown)):
(TestWebKitAPI::TEST(SiteIsolation, ResponsiveProcessAfterMousedown)):
(TestWebKitAPI::TEST(SiteIsolation, UnresponsiveProcessMousedown)):
(TestWebKitAPI::TEST(SiteIsolation, UnresponsiveProcessDies)):

Canonical link: <a href="https://commits.webkit.org/303876@main">https://commits.webkit.org/303876@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57f1c13c205e666d3059791ee26be9e790ca25e1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133831 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6341 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45020 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141408 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5f767642-e581-4446-a4e1-4bf3fe23da67) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135701 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6870 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6206 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102376 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/e95c0df8-1dfa-42d8-8633-1e411a987ca8) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136778 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119974 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83174 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1d72810b-45cb-45ee-b611-9368a7f8024f) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2342 "Passed tests") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/113878 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38091 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144054 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6012 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38671 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110747 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6094 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5131 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110942 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4577 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116228 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59759 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20686 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6064 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34514 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5909 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69528 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6155 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6018 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -11537,9 +11548,14 @@ void WebPageProxy::processDidBecomeUnresponsive()
         return;
 
+    bool wasResponsive = !m_unresponsiveProcessCount;
+    ++m_unresponsiveProcessCount;
```
Don't we need to decrease `m_unresponsiveProcessCount` when some unresponsive process is killed or no longer used by the web page?

rniwa replied:
oh, that's a good point! maybe we should use WeakHashSet instead.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -11253,7 +11264,7 @@ void WebPageProxy::didReceiveEvent(IPC::Connection* connection, WebEventType eve
     case WebEventType::GestureEnd:
 #endif
-        protectedLegacyMainFrameProcess()->stopResponsivenessTimer();
+        WebProcessProxy::fromConnection(*connection)->stopResponsivenessTimer();
```
Connection does need to be null checked.  If WebPageProxy::sendGestureEvent fails, it will call this with a null connection.

rniwa replied:
Sure, added.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -11530,28 +11541,38 @@ void WebPageProxy::focusedFrameChanged(IPC::Connection& connection, std::optiona
     if (!hasRunningProcess())
         return;
 
+    bool wasResponsive = m_unresponsiveProcesses.isEmpty();
```
This is not necessarily true if m_unresponsiveProcesses contained only other processes.

rniwa replied:
? If there are other unresponsive processes, then we weren't responsive?

bnham replied:
I also don't completely understand the comment. Maybe a concrete example would be useful?

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -11530,28 +11541,38 @@ void WebPageProxy::focusedFrameChanged(IPC::Connection& connection, std::optiona
         return;
 
+    m_unresponsiveProcesses.remove(process);
+    bool isReponsive = m_unresponsiveProcesses.isEmpty();
```
Ditto

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -4043,6 +4043,8 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
     bool m_menuBarIsVisible { true };
     bool m_toolbarsAreVisible { true };
+
+    HashSet<CheckedRef<WebProcessProxy>> m_unresponsiveProcesses;
```
I'm not convinced there is no way for a process to be destroyed while still being in this set.  Could we use weak pointers instead of checked?

rniwa replied:
If we did that, then WebPageProxy can become responsive without notifying the client, and I don't think we want that.

bnham reviewed (approved):
Seems reasonable to me.

szewai commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -11582,28 +11594,38 @@ void WebPageProxy::focusedFrameChanged(IPC::Connection& connection, std::optiona
     if (!hasRunningProcess())
         return;
 
+    m_unresponsiveProcesses.remove(process);
```
We probably should notify `processDidBecomeResponsive` only if the target process is in `m_unresponsiveProcesses`?

rniwa replied:
I don't think `WebPageProxy::processDidBecomeResponsive` will be called unless it was in `m_unresponsiveProcesses`. I can add a debug ASSERT though.,

rniwa committed:
[Site Isolation] Make unresponsive process handling work
https://bugs.webkit.org/show_bug.cgi?id=302842

Reviewed by Sihui Liu and Ben Nham.

Make the handling of foreground unresponsiveness work with site isolation.

Test: TestWebKitAPI.SiteIsolation.UnresponsiveProcessKeydown
      TestWebKitAPI.SiteIsolation.ResponsiveProcessAfterMousedown
      TestWebKitAPI.SiteIsolation.UnresponsiveProcessMousedown
      TestWebKitAPI.SiteIsolation.UnresponsiveProcessDies

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::startResponsivenessTimerForMouseEvent):
(WebKit::WebPageProxy::processNextQueuedMouseEvent):
(WebKit::WebPageProxy::processNextQueuedGestureEvent):
(WebKit::WebPageProxy::sendKeyEvent):
(WebKit::WebPageProxy::handleKeyboardEvent):
(WebKit::WebPageProxy::mouseEventHandlingCompleted):
(WebKit::WebPageProxy::didReceiveEvent):
(WebKit::WebPageProxy::processDidBecomeUnresponsive):
(WebKit::WebPageProxy::processDidBecomeResponsive):
(WebKit::WebPageProxy::dispatchProcessDidTerminate):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::removeRemotePageProxy):
(WebKit::WebProcessProxy::removeWebPage):
(WebKit::WebProcessProxy::didBecomeUnresponsive):
(WebKit::WebProcessProxy::didBecomeResponsive):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(-[NavigationDelegateWithUnresponsiveCallback webView:didReceiveAuthenticationChallenge:completionHandler:]):
(-[NavigationDelegateWithUnresponsiveCallback waitForDidFinishNavigation]):
(-[NavigationDelegateWithUnresponsiveCallback didBecomeUnresponsive]):
(-[NavigationDelegateWithUnresponsiveCallback didBecomeResponsive]):
(-[NavigationDelegateWithUnresponsiveCallback _webViewWebProcessDidBecomeUnresponsive:]):
(-[NavigationDelegateWithUnresponsiveCallback _webViewWebProcessDidBecomeResponsive:]):
(-[NavigationDelegateWithUnresponsiveCallback webView:didFinishNavigation:]):
(TestWebKitAPI::TEST(SiteIsolation, UnresponsiveProcessKeydown)):
(TestWebKitAPI::TEST(SiteIsolation, ResponsiveProcessAfterMousedown)):
(TestWebKitAPI::TEST(SiteIsolation, UnresponsiveProcessMousedown)):
(TestWebKitAPI::TEST(SiteIsolation, UnresponsiveProcessDies)):

Canonical link: https://commits.webkit.org/303876@main
